sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -414,8 +414,42 @@ class Analyzer(
     }
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-      case i @ InsertIntoTable(u: UnresolvedRelation, _, _, _, _) =>
-        i.copy(table = EliminateSubqueryAliases(getTable(u)))
+      case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        val table = getTable(u)
+        // adding the table's partitions or validate the query's partition info
+        table match {
+          case relation: PartitionedRelation if relation.partitionColumns.nonEmpty =>
+            val tablePartitionNames = relation.partitionColumns.map(_.name)
+            if (parts.keys.nonEmpty) {
+              // the query's partitioning must match the table's partitioning
+              // this is set for queries like: insert into ... partition (one = "a", two = <expr>)
+              if (tablePartitionNames.size != parts.keySet.size) {
```

Comment:
Removing this check broke 2 tests because the pre-insert code isn't correctly checking it. My follow-up PR fixes and merges the pre-inserts, so I've added it back in this PR with a TODO to remove it. I'll remove it when I rebase the follow-up on this.